File: lib/core/visibility_rules.dart
/// Kimin hangi aracı göreceğini belirleyen kurallar.
///
/// Tamamı saf fonksiyondur: Firebase, widget ya da zaman bağımlılığı yoktur,
/// doğrudan birim testle kaplanır.
///
/// Not: bu kurallar bir **arayüz** filtresidir, güvenlik sınırı değildir.
/// Veritabanı kuralları `vehicles` düğümünü giriş yapmış herkese okutur.
library;

import '../models/group_config.dart';
import '../models/vehicle.dart';

/// Bu kullanıcıya görünecek araçlar. Giriş sırası korunur.
///
/// * Yönetici: tüm araçlar.
/// * Onaysız ya da grupsuz sürücü: yalnızca kendi aracı.
/// * Onaylı ve gruplu sürücü: kendi aracı + kendi grubu + grubunun
///   `visibleGroups` listesindeki gruplar. Kendisi dışındaki onaysız
///   araçlar gösterilmez.
List<Vehicle> visibleVehiclesFor({
  required List<Vehicle> allVehicles,
  required List<GroupConfig> groups,
  required String viewerUid,
  required bool isAdmin,
}) {
  if (isAdmin) return List<Vehicle>.of(allVehicles);

  final Vehicle? viewer = findVehicleById(allVehicles, viewerUid);
  // Kaydı olmayan sürücü henüz hiçbir şey görmez.
  if (viewer == null) return const <Vehicle>[];

  // Kendi aracı her zaman görünür, onaysız ve grupsuz olsa bile.
  if (!viewer.approved || !viewer.hasGroup) return <Vehicle>[viewer];

  final Set<String> allowed = allowedGroupIdsFor(
    viewerGroupId: viewer.groupId!,
    groups: groups,
  );

  return <Vehicle>[
    for (final Vehicle vehicle in allVehicles)
      if (vehicle.id == viewerUid ||
          (vehicle.approved &&
              vehicle.hasGroup &&
              allowed.contains(vehicle.groupId)))
        vehicle,
  ];
}

/// Sürücünün görebileceği grup kimlikleri: kendi grubu + ayarındaki gruplar.
Set<String> allowedGroupIdsFor({
  required String viewerGroupId,
  required List<GroupConfig> groups,
}) {
  final Set<String> allowed = <String>{viewerGroupId};
  final GroupConfig? config = findGroupById(groups, viewerGroupId);
  if (config != null) allowed.addAll(config.visibleGroups);
  return allowed;
}

/// Bir aracın hangi alanlarının gösterileceği.
class FieldVisibility {
  const FieldVisibility({
    required this.showSpeed,
    required this.showDriverName,
  });

  static const FieldVisibility full = FieldVisibility(
    showSpeed: true,
    showDriverName: true,
  );

  final bool showSpeed;
  final bool showDriverName;

  @override
  bool operator ==(Object other) {
    return other is FieldVisibility &&
        other.showSpeed == showSpeed &&
        other.showDriverName == showDriverName;
  }

  @override
  int get hashCode => Object.hash(showSpeed, showDriverName);

  @override
  String toString() =>
      'FieldVisibility(hız: $showSpeed, sürücü adı: $showDriverName)';
}

/// Alan gizleme, **bakılan aracın** grup ayarına göre yapılır: bir grup
/// kendi araçlarının hızını ya da sürücü adını başkalarından gizleyebilir.
///
/// İki istisna tam görünürlük verir:
/// * kullanıcı kendi aracına bakıyorsa,
/// * yöneticiye bakıyorsa (filoyu yönetebilmesi için).
FieldVisibility fieldVisibilityFor({
  required Vehicle vehicle,
  required List<GroupConfig> groups,
  required String viewerUid,
  required bool isAdmin,
}) {
  if (vehicle.id == viewerUid) return FieldVisibility.full;
  if (isAdmin) return FieldVisibility.full;
  if (!vehicle.hasGroup) return FieldVisibility.full;

  final GroupConfig? config = findGroupById(groups, vehicle.groupId!);
  if (config == null) return FieldVisibility.full;
  return FieldVisibility(
    showSpeed: config.showSpeed,
    showDriverName: config.showDriverName,
  );
}

/// Yönetici haritasındaki grup filtresi çipleri.
enum GroupFilterKind { all, ungrouped, specific }

class GroupFilter {
  const GroupFilter.all() : kind = GroupFilterKind.all, groupId = null;

  const GroupFilter.ungrouped()
    : kind = GroupFilterKind.ungrouped,
      groupId = null;

  const GroupFilter.group(String this.groupId) : kind = GroupFilterKind.specific;

  final GroupFilterKind kind;
  final String? groupId;

  bool matches(Vehicle vehicle) {
    switch (kind) {
      case GroupFilterKind.all:
        return true;
      case GroupFilterKind.ungrouped:
        return !vehicle.hasGroup;
      case GroupFilterKind.specific:
        return vehicle.groupId == groupId;
    }
  }

  @override
  bool operator ==(Object other) {
    return other is GroupFilter &&
        other.kind == kind &&
        other.groupId == groupId;
  }

  @override
  int get hashCode => Object.hash(kind, groupId);
}

Vehicle? findVehicleById(List<Vehicle> vehicles, String id) {
  for (final Vehicle vehicle in vehicles) {
    if (vehicle.id == id) return vehicle;
  }
  return null;
}

GroupConfig? findGroupById(List<GroupConfig> groups, String groupId) {
  for (final GroupConfig group in groups) {
    if (group.groupId == groupId) return group;
  }
  return null;
}

File: panel/js/visibility-rules.js
// Kimin hangi aracı göreceğini belirleyen kurallar.
//
// lib/core/visibility_rules.dart dosyasının panel karşılığıdır. Uygulamada
// "bakan kişi" her zaman bir araçtır (sürücünün uid'i aynı zamanda aracın
// kimliğidir); panelde ise bakan kişi araçsız bir hesaptır, grubu
// `/webUsers/{uid}.groupId` alanından gelir. Grup çözümü ve alan gizleme
// mantığı ikisinde de aynıdır.
//
// Not: bu kurallar bir **arayüz** filtresidir, güvenlik sınırı değildir.
// Veritabanı kuralları `vehicles` düğümünü giriş yapmış herkese okutur.

import { Role } from './models.js';

// Bakanın görebileceği grup kimlikleri: kendi grubu + ayarındaki gruplar.
export function allowedGroupIdsFor(viewerGroupId, groups) {
  const allowed = new Set([viewerGroupId]);
  const config = findGroupById(groups, viewerGroupId);
  if (config !== null) for (const id of config.visibleGroups) allowed.add(id);
  return allowed;
}

// Bu kullanıcıya görünecek araçlar. Giriş sırası korunur.
//
// * Yönetici: tüm araçlar.
// * İzleyici: kendi grubu + grubunun visibleGroups listesindeki grupların
//   **onaylı** araçları. Onay bekleyen araçlar yalnızca yöneticiye görünür.
// * Grubu olmayan izleyici ve onay bekleyen hesap: hiçbir araç.
export function visibleVehiclesFor(allVehicles, groups, viewer) {
  if (viewer.role === Role.admin) return [...allVehicles];
  if (viewer.role !== Role.viewer) return [];

  const viewerGroupId = viewer.groupId;
  if (viewerGroupId === null || viewerGroupId.length === 0) return [];

  const allowed = allowedGroupIdsFor(viewerGroupId, groups);
  return allVehicles.filter(
    (vehicle) => vehicle.approved && vehicle.hasGroup && allowed.has(vehicle.groupId),
  );
}

// Bir aracın hangi alanlarının gösterileceği.
const FULL_VISIBILITY = Object.freeze({ showSpeed: true, showDriverName: true });

// Alan gizleme, **bakılan aracın** grup ayarına göre yapılır: bir grup kendi
// araçlarının hızını ya da sürücü adını başkalarından gizleyebilir.
//
// Yönetici filoyu yönetebilmesi için her zaman tam görür.
export function fieldVisibilityFor(vehicle, groups, viewer) {
  if (viewer.role === Role.admin) return FULL_VISIBILITY;
  if (!vehicle.hasGroup) return FULL_VISIBILITY;

  const config = findGroupById(groups, vehicle.groupId);
  if (config === null) return FULL_VISIBILITY;
  return { showSpeed: config.showSpeed, showDriverName: config.showDriverName };
}

// Haritadaki grup filtresi çipleri.
export const GroupFilterKind = Object.freeze({
  all: 'all',
  ungrouped: 'ungrouped',
  specific: 'specific',
});

export function groupFilterAll() {
  return { kind: GroupFilterKind.all, groupId: null };
}

export function groupFilterUngrouped() {
  return { kind: GroupFilterKind.ungrouped, groupId: null };
}

export function groupFilterFor(groupId) {
  return { kind: GroupFilterKind.specific, groupId };
}

export function filterMatches(filter, vehicle) {
  switch (filter.kind) {
    case GroupFilterKind.all:
      return true;
    case GroupFilterKind.ungrouped:
      return !vehicle.hasGroup;
    case GroupFilterKind.specific:
      return vehicle.groupId === filter.groupId;
    default:
      return true;
  }
}

export function filtersEqual(a, b) {
  return a.kind === b.kind && a.groupId === b.groupId;
}

export function findGroupById(groups, groupId) {
  return groups.find((group) => group.groupId === groupId) ?? null;
}
